### ISSUE TITLE
invalid memory address or nil pointer dereference

### ISSUE THREAD

**Issue by erebusdark** [kpango/fastime#31]
I am using the github.com/kpango/glg package that in turn is using this package. When I run my code on a x86_64 machine it works, no problems what so ever, but when I compile it and try to run it on a arm processor (Raspberry Pi 3) it panics and crashes with the following error

`panic: runtime error: invalid memory address or nil pointer dereference`
`[signal SIGSEGV: segmentation violation code=0x1 addr=0x4 pc=0x114fc]`
`goroutine 17 [running]:`
`sync/atomic.storeUint64(0x1101e034, 0x4c4b40, 0x0)`
`	/usr/lib/go-1.10/src/sync/atomic/64bit_arm.go:20 +0x3c`
`github.com/kpango/fastime.(*Fastime).StartTimerD.func1(0x1101e000, 0x4c4b40, 0x0, 0x386220, 0x110001e0)`
`	/home/user/go/src/github.com/kpango/fastime/fastime.go:181 +0x64`
`created by github.com/kpango/fastime.(*Fastime).StartTimerD`
`	/home/user/go/src/github.com/kpango/fastime/fastime.go:179 +0xb8`

**Comment by kpango** [comment]
@erebusdark Hi. I'm so sorry I didn't notice about this issue.

which Go version are u using? 1.10?

**Comment by kpango** [comment]
@erebusdark can u test on Go 1.13?

**Comment by erebusdark** [comment]
I will check with Go 1.13 during next week when I am back so I can test it. Will update this once I have tested it.

**Comment by erebusdark** [comment]
I was using 1.10 but get the same error with 1.13.1

**Comment by kpango** [comment]
can you show me the error log with Go 1.13.1?

**Comment by erebusdark** [comment]
```panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x4 pc=0x12488]

goroutine 18 [running]:
github.com/kpango/fastime.(*Fastime).StartTimerD.func1(0x1c84140, 0x4c4b40, 0x0, 0x2bf3e8, 0x1c6a9a0)
        /home/erebus/go/src/github.com/kpango/fastime/fastime.go:181 +0x60
created by github.com/kpango/fastime.(*Fastime).StartTimerD
        /home/erebus/go/src/github.com/kpango/fastime/fastime.go:179 +0xb4
```

The file is very simple and it looks like this:
```
package main

import (
	"github.com/kpango/glg"
)

func main() {
	infolog := glg.FileWriter("./info.log", 0666)
	errorlog := glg.FileWriter("./error.log", 0666)

	defer infolog.Close()
	defer errorlog.Close()

	glg.Get().
		SetMode(glg.BOTH).
		AddLevelWriter(glg.ERR, errorlog).
		AddLevelWriter(glg.INFO, infolog)
}
```

**Comment by kpango** [comment]
could you please give me raw. stdout result?
below is too short

```
[signal SIGSEGV: segmentation violation code=0x1 addr=0x4 pc=0x12488]

goroutine 18 [running]:
github.com/kpango/fastime.(*Fastime).StartTimerD.func1(0x1c84140, 0x4c4b40, 0x0, 0x2bf3e8, 0x1c6a9a0)
        /home/erebus/go/src/github.com/kpango/fastime/fastime.go:181 +0x60
created by github.com/kpango/fastime.(*Fastime).StartTimerD
        /home/erebus/go/src/github.com/kpango/fastime/fastime.go:179 +0xb4

```

**Comment by erebusdark** [comment]
I am not sure how to get the raw stdout.

**Comment by kpango** [comment]
please paste raw result below
```shell
echo "$(go version)\n$(go run main.go)" >> result.txt
```
the code of fastime.go:181, I just use atomic.StoreUint64() there is no problem about that we can find stdlibs problem by assembly error like you paste
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x4 pc=0x114fc]
goroutine 17 [running]:
sync/atomic.storeUint64(0x1101e034, 0x4c4b40, 0x0)
/usr/lib/go-1.10/src/sync/atomic/64bit_arm.go:20 +0x3c
github.com/kpango/fastime.(*Fastime).StartTimerD.func1(0x1101e000, 0x4c4b40, 0x0, 0x386220, 0x110001e0)
/home/user/go/src/github.com/kpango/fastime/fastime.go:181 +0x64
created by github.com/kpango/fastime.(*Fastime).StartTimerD
/home/user/go/src/github.com/kpango/fastime/fastime.go:179 +0xb8
```

**Comment by erebusdark** [comment]
Ah, that is a problem sense I don't have go on the raspberry pi.
I compile it with "env GOOS=linux GOARCH=arm GOARM=7 go build" and then move the executable to the pi.

I did just find this however:

> On ARM, x86-32, and 32-bit MIPS, it is the caller's responsibility to arrange for 64-bit alignment of 64-bit words accessed atomically. The first word in a variable or in an allocated struct, array, or slice can be relied upon to be 64-bit aligned.

from https://golang.org/pkg/sync/atomic/
so I am guessing that has something to do with it.

**Comment by kpango** [comment]
Oh that's make sense.
So, I should pre-allocate f.dur memory not initial value.
Please use master version of fastime I implemented pre-allocate things.
https://github.com/kpango/fastime/commit/c67c73ee0febbb6fd7685ca5ab5aaa9e52496d2b

**Comment by erebusdark** [comment]
I tried it with the latest commit and the small test program I wrote (shown above) runs without crashing now so I believe it to be solved now.
Excellent work (both on the package and fixing the bug/issue)

**Comment by kpango** [comment]
hmm :thinking: 
So you mean it crashed again at the same line of code?

**Comment by erebusdark** [comment]
Yes, and with the same error message.

**Comment by kpango** [comment]
@erebusdark Hi, I apologize that this issue has been lying unsolved for a very long time.
I have just released fastime v1.0.17 and glg v1.6.2 which uses it, since the alignment issue was solved by #42.

You might not have already used them, but I'd like to give you a reminder.
